src/workerd/io/actor-cache.h

```diff
@@ -573,6 +583,35 @@ class ActorCache final: public ActorCacheInterface {
 
   kj::HashSet<CountedDelete*> countedDeletes;
 
+  // A convenient object for holding all our `Entry`s that are waiting on the result of a get() or
+  // getMultiple() call. `get()` will only ever have 1 element in `entries`, unlike `getMultiple()`.
+  class GetWaiter {
+  public:
+    explicit GetWaiter(ActorCache& cache, kj::Array<kj::Own<Entry>> entries)
+      : cache(cache), entries(kj::mv(entries)) {}
+    KJ_DISALLOW_COPY_AND_MOVE(GetWaiter);
+    ~GetWaiter() noexcept(false) {
+      for (auto& entry : entries) {
+        if (!entry->isShared()) {
```

Comment:
This "cancel if the `GetWaiter` is destroyed" approach doesn't really work as is now. 

If the `GetWaiter` is holding the last reference to the `Entry`, then upon its destruction (because JS dropped the get() promise) we would remove the `Entry` from the `dirtyList` and we wouldn't attempt to do the get() when we get around to flushing. The thing is, we almost always give the `currentValues` map a strong reference to the `Entry`, then `addRef` that and put it in the `dirtyList` (see `findInCache()`). This means by the time the `GetWaiter` has been initialized, we will (in almost all cases) have at least 2 strong references to the `Entry` -- one owned by the `GetWaiter` and one owned by `currentValues`. When `GetWaiter`'s dtor runs, we won't even try to remove the Entry from `dirtyList` because it looks like it's shared.

I think one way to get around this is to change the check from "is shared" to "our Entry matches what's in `currentValues` + the status is `UNKNOWN`". This would imply we haven't changed the Entry since we initiated the `get()`.

---
There's an open question on if we want to try to cancel the storage get() during the flush. I think this might be more complicated than it's worth, since we would have already put `Entry`s into a `GetFlush`, and the storage operation might have already run. What if the values have already been retrieved from storage, and the waiter is just waiting for the flush promise to finish?

Reply:
> I think one way to get around this is to change the check from "is shared" to "our Entry matches what's in `currentValues` + the status is `UNKNOWN`". This would imply we haven't changed the Entry since we initiated the `get()`.

But what if some other `get()` request has been made for the same key? Couldn't that logic cause a false positive that would inadvertently cancel the read when it shouldn't?

(again, sorry for commenting on this without having yet read the full implementation -- I've gone from the test file to the header file so far).

---

> There's an open question on if we want to try to cancel the storage get() during the flush.

If the storage operation has already been sent (or packed into a `GetFlush` in preparation for sending), then yeah I see little marginal return in trying to cancel it. But it does (to me at least) seem meaningfully more valuable to attempt to avoid sending a get to storage if it's canceled before we get around to the flush call.

Reply:
> But what if some other get() request has been made for the same key? Couldn't that logic cause a false positive that would inadvertently cancel the read when it shouldn't?

Yeah, that's a solid point. If we have 2 outstanding gets() for one key (i.e. two separate `GetWaiter`s), then they'll each hold a strong ref (via `addAtomicRef()`), and so if one of the `GetWaiter`s is destroyed, we still can't assume with that a match within `currentValue` & status `UNKNOWN` means we're the only outstanding get.

Maybe we can put a counter on the `Entry` which we increment/decrement whenever a GetWaiter with that Entry is created/destroyed? If it goes down to 1 we know we're the last waiter?